service_handler.go

```diff
@@ -144,22 +146,22 @@ func AddressBookNew(session Session) *AddressBook {
 
 // AddService adds a service
 func (s *AddressBook) AddService(service *Service) {
-	s.Mutex.Lock()
+	s.mutex.Lock()
 	rawurl := service.URL.String()
 	if entry, ok := s.entries[rawurl]; ok {
 		if entry.running {
 			entry.Stop()
 		}
 	}
 	s.entries[rawurl] = &*service
-	s.Mutex.Unlock()
+	s.mutex.Unlock()
```

Comment:
I recommend using `defer s.mutex.Unlock()` right after you lock the mutex to make sure you don't forget to unlock it in all code paths.

Reply:
fixed